## **EDA**  exploratory data analysis

**FASE 1**

In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_columns', None)

#### ENTENDER LOS DATOS

In [3]:
df = pd.read_csv("hr_raw_data.csv", index_col = 0)

df.head(2)

age attrition businesstravel    dailyrate department  distancefromhome  \
0  51        No            NaN  2015.722222        NaN                 6   
1  52        No            NaN  2063.388889        NaN                 1   

   education educationfield  employeecount  employeenumber  \
0          3            NaN              1               1   
1          4  Life Sciences              1               2   

   environmentsatisfaction  gender  hourlyrate  jobinvolvement  joblevel  \
0                        1       0         NaN               3         5   
1                        3       0         NaN               2         5   

               jobrole  jobsatisfaction maritalstatus monthlyincome  \
0   resEArch DIREcToR                 3           NaN     16280,83$   
1             ManAGeR                 3           NaN           NaN   

  monthlyrate  numcompaniesworked over18 overtime  percentsalaryhike  \
0   42330,17$                   7      Y       No                 13   
1   43331,17$                   0    NaN      NaN                 14   

  performancerating  relationshipsatisfaction standardhours  stockoptionlevel  \
0               3,0                         3     Full Time                 0   
1               3,0                         1           NaN                 1   

  totalworkingyears  trainingtimeslastyear worklifebalance  yearsatcompany  \
0               NaN                      5             3,0              20   
1              34,0                      5             3,0              33   

  yearsincurrentrole  yearssincelastpromotion  yearswithcurrmanager  \
0                NaN                       15                    15   
1                NaN                       11                     9   

  sameasmonthlyincome  datebirth      salary roledepartament  numberchildren  \
0           16280,83$       1972  195370,00$             NaN             NaN   
1                 NaN       1971  199990,00$             NaN             NaN   

  remotework  
0        Yes  
1          1

In [5]:
print(f"El número de filas que tenemos es {df.shape[0]}, y el número de columnas es {df.shape[1]}")

El número de filas que tenemos es 1678, y el número de columnas es 41


In [6]:
# TIPOS DE DATOS

df.dtypes

age                          object
attrition                    object
businesstravel               object
dailyrate                   float64
department                   object
distancefromhome              int64
education                     int64
educationfield               object
employeecount                 int64
employeenumber                int64
environmentsatisfaction       int64
gender                        int64
hourlyrate                  float64
jobinvolvement                int64
joblevel                      int64
jobrole                      object
jobsatisfaction               int64
maritalstatus                object
monthlyincome                object
monthlyrate                  object
numcompaniesworked            int64
over18                       object
overtime                     object
percentsalaryhike             int64
performancerating            object
relationshipsatisfaction      int64
standardhours                object
stockoptionlevel            

In [7]:
# PORCENTAJE DE NULOS

round(df.isnull().sum()/df.shape[0]*100, 2)

age                           0.00
attrition                     0.00
businesstravel               47.74
dailyrate                     0.00
department                   81.41
distancefromhome              0.00
education                     0.00
educationfield               46.13
employeecount                 0.00
employeenumber                0.00
environmentsatisfaction       0.00
gender                        0.00
hourlyrate                   75.51
jobinvolvement                0.00
joblevel                      0.00
jobrole                       0.00
jobsatisfaction               0.00
maritalstatus                40.23
monthlyincome                29.14
monthlyrate                   0.00
numcompaniesworked            0.00
over18                       55.90
overtime                     41.48
percentsalaryhike             0.00
performancerating            11.92
relationshipsatisfaction      0.00
standardhours                20.92
stockoptionlevel              0.00
totalworkingyears   

In [ ]:
#VER CÓMO SON LOS DATOS QUE TENEMOS
df.describe().T

count         mean         std          min  \
dailyrate                 1678.0   668.079714  470.787298   104.103175   
distancefromhome          1678.0     4.504172   14.652066   -49.000000   
education                 1678.0     2.932658    1.024270     1.000000   
employeecount             1678.0     1.000000    0.000000     1.000000   
employeenumber            1678.0   809.859952  467.084867     1.000000   
environmentsatisfaction   1678.0     4.264005    6.912695     1.000000   
gender                    1678.0     0.398689    0.489774     0.000000   
hourlyrate                 411.0    83.140768   57.272101    13.012897   
jobinvolvement            1678.0     2.740763    0.710359     1.000000   
joblevel                  1678.0     2.064362    1.099425     1.000000   
jobsatisfaction           1678.0     2.743147    1.105172     1.000000   
numcompaniesworked        1678.0     2.670441    2.501133     0.000000   
percentsalaryhike         1678.0    15.154350    3.649607    11.000000   
relationshipsatisfaction  1678.0     2.700238    1.079494     1.000000   
stockoptionlevel          1678.0     0.787843    0.843252     0.000000   
trainingtimeslastyear     1678.0     2.795590    1.288657     0.000000   
yearsatcompany            1678.0     7.124553    6.173837     0.000000   
yearssincelastpromotion   1678.0     2.241955    3.246554     0.000000   
yearswithcurrmanager      1678.0     4.200238    3.554224     0.000000   
datebirth                 1678.0  1986.047080    9.154908  1963.000000   
numberchildren               0.0          NaN         NaN          NaN   

                                  25%          50%          75%          max  
dailyrate                  290.035510   556.256661   971.956349  2063.388889  
distancefromhome             2.000000     5.000000    11.000000    29.000000  
education                    2.000000     3.000000     4.000000     5.000000  
employeecount                1.000000     1.000000     1.000000     1.000000  
employeenumber             403.250000   813.500000  1215.750000  1614.000000  
environmentsatisfaction      2.000000     3.000000     4.000000    49.000000  
gender                       0.000000     0.000000     1.000000     1.000000  
hourlyrate                  36.254439    69.532083   116.987103   255.963294  
jobinvolvement               2.000000     3.000000     3.000000     4.000000  
joblevel                     1.000000     2.000000     3.000000     5.000000  
jobsatisfaction              2.000000     3.000000     4.000000     4.000000  
numcompaniesworked           1.000000     2.000000     4.000000     9.000000  
percentsalaryhike           12.000000    14.000000    18.000000    25.000000  
relationshipsatisfaction     2.000000     3.000000     4.000000     4.000000  
stockoptionlevel             0.000000     1.000000     1.000000     3.000000  
trainingtimeslastyear        2.000000     3.000000     3.000000     6.000000  
yearsatcompany               3.000000     5.000000     9.000000    40.000000  
yearssincelastpromotion      0.000000     1.000000     3.000000    15.000000  
yearswithcurrmanager         2.000000     3.000000     7.000000    17.000000  
datebirth                 1980.000000  1987.000000  1993.000000  2005.000000  
numberchildren                    NaN          NaN          NaN          NaN

In [6]:
#VER CUÁLES SON TIPO OBJECT
df.select_dtypes(include = "object")


age attrition businesstravel                department    educationfield  \
0     51        No            NaN                       NaN               NaN   
1     52        No            NaN                       NaN     Life Sciences   
2     42        No  travel_rarely   Research & Development   Technical Degree   
3     47        No  travel_rarely                       NaN           Medical   
4     46        No            NaN                       NaN  Technical Degree   
...   ..       ...            ...                       ...               ...   
1673  43        No            NaN                       NaN           Medical   
1674  47        No            NaN                       NaN               NaN   
1675  29        No  travel_rarely                       NaN               NaN   
1676  47        No  travel_rarely                       NaN     Life Sciences   
1677  32        No            NaN                       NaN     Life Sciences   

                       jobrole maritalstatus monthlyincome monthlyrate over18  \
0           resEArch DIREcToR            NaN     16280,83$   42330,17$      Y   
1                     ManAGeR            NaN           NaN   43331,17$    NaN   
2                     ManaGER        Married           NaN   41669,33$    NaN   
3           ReseArCH DIrECtOr        Married     14307,50$   37199,50$      Y   
4             sAleS EXECUtIve       Divorced     12783,92$   33238,20$      Y   
...                        ...           ...           ...         ...    ...   
1673       rESEaRcH SciEnTiST         Single      3949,17$   10267,83$    NaN   
1674                  mANager        Married     15943,72$   41453,67$      Y   
1675       reSearch sCienTiSt            NaN           NaN    6090,75$    NaN   
1676   maNufACTURING DIREctOr       Divorced      8339,32$   21682,23$    NaN   
1677       resEArch scIENTisT         Single           NaN   11681,39$      Y   

     overtime performancerating standardhours totalworkingyears  \
0          No               3,0     Full Time               NaN   
1         NaN               3,0           NaN              34,0   
2          No               3,0           NaN              22,0   
3         NaN               3,0     Full Time               NaN   
4          No               3,0           NaN               NaN   
...       ...               ...           ...               ...   
1673      NaN               3,0           NaN               NaN   
1674       No               3,0     Full Time              27,0   
1675       No               3,0     Part Time               6,0   
1676      Yes               NaN     Part Time               NaN   
1677      Yes               3,0     Part Time              10,0   

     worklifebalance yearsincurrentrole sameasmonthlyincome      salary  \
0                3,0                NaN           16280,83$  195370,00$   
1                3,0                NaN                 NaN  199990,00$   
2                NaN                NaN                 NaN  192320,00$   
3                NaN                NaN           14307,50$  171690,00$   
4                3,0                NaN           12783,92$         NaN   
...              ...                ...                 ...         ...   
1673             3,0                NaN            3949,17$         NaN   
1674             3,0                NaN           15943,72$  191324,62$   
1675             3,0                NaN                 NaN   28111,13$   
1676             3,0                NaN            8339,32$  100071,84$   
1677             3,0                NaN                 NaN   53914,11$   

                           roledepartament remotework  
0                                      NaN        Yes  
1                                      NaN          1  
2      ManaGER  -  Research & Development           1  
3                                      NaN      False  
4                                      NaN          0  
...              

In [8]:
df['businesstravel'].value_counts()

businesstravel
travel_rarely        616
travel_frequently    168
non-travel            93
Name: count, dtype: int64

In [9]:
# ENTENDER LA RELACIÓN Y DIFERENCIAS ENTRE ESTAS COLUMNAS

df[['jobrole', 'dailyrate', 'monthlyincome', 'monthlyrate', 'salary']]

jobrole    dailyrate monthlyincome monthlyrate  \
0           resEArch DIREcToR   2015.722222     16280,83$   42330,17$   
1                     ManAGeR   2063.388889           NaN   43331,17$   
2                     ManaGER   1984.253968           NaN   41669,33$   
3           ReseArCH DIrECtOr   1771.404762     14307,50$   37199,50$   
4             sAleS EXECUtIve   1582.771346     12783,92$   33238,20$   
...                        ...          ...           ...         ...   
1673       rESEaRcH SciEnTiST    488.944444      3949,17$   10267,83$   
1674                  mANager   1973.984127     15943,72$   41453,67$   
1675       reSearch sCienTiSt    290.035510           NaN    6090,75$   
1676   maNufACTURING DIREctOr   1032.487286      8339,32$   21682,23$   
1677       resEArch scIENTisT    556.256661           NaN   11681,39$   

          salary  
0     195370,00$  
1     199990,00$  
2     192320,00$  
3     171690,00$  
4            NaN  
...          ...  
1673         NaN  
1674  191324,62$  
1675   28111,13$  
1676  100071,84$  
1677   53914,11$  

[1678 rows x 5 columns]

In [10]:
df['department'].value_counts()

department
Research & Development     203
Sales                       93
Human Resources             16
Name: count, dtype: int64

In [11]:
df[['distancefromhome']].value_counts()

distancefromhome
 2                  228
 1                  209
 9                   89
 10                  86
 8                   85
                   ... 
-34                   2
-28                   2
-21                   2
-40                   1
-39                   1
Name: count, Length: 69, dtype: int64

In [12]:
df['education'].value_counts()

education
3    649
4    461
2    322
1    186
5     60
Name: count, dtype: int64

In [13]:
df['educationfield'].value_counts()

educationfield
Life Sciences       367
Medical             286
Marketing           106
Technical Degree     70
Other                63
Human Resources      12
Name: count, dtype: int64

In [14]:
df['employeecount'].value_counts()

employeecount
1    1678
Name: count, dtype: int64

In [15]:
df['employeenumber'].duplicated().sum()

np.int64(64)

In [16]:
df['environmentsatisfaction'].unique()

array([ 1,  3,  4,  2, 42, 37, 35, 25, 27, 31, 39, 21, 15, 14, 33, 19, 12,
       13, 28, 47, 36, 29, 24, 46, 16, 22, 41, 49, 11, 48, 18, 10, 45, 38,
       17, 20, 26, 43])

In [17]:
df['relationshipsatisfaction'].value_counts()

relationshipsatisfaction
3    527
4    483
2    350
1    318
Name: count, dtype: int64

In [18]:
df["relationshipsatisfaction"].mean()


np.float64(2.700238379022646)

In [19]:
df['standardhours'].value_counts()

standardhours
Part Time    927
Full Time    400
Name: count, dtype: int64

In [20]:
df['stockoptionlevel'].value_counts()

stockoptionlevel
0    719
1    689
2    177
3     93
Name: count, dtype: int64

**stock options**
- son incentivos que permiten a los empleados comprar acciones de la empresa a un precio fijo (generalmente más bajo que el de mercado) después de cierto tiempo o cumpliendo ciertos objetivos.

- es un indicador de cuánto puede beneficiarse un empleado de las stock options dentro de la empresa.



    0 → No tiene asignadas stock options.

    1 → Nivel básico, pocas opciones asignadas.

    2 → Nivel intermedio, más opciones que el nivel 1.

    3 → Nivel alto, la mayor cantidad de opciones asignadas, normalmente para puestos clave o ejecutivos.

En otras palabras, a mayor número, mayor beneficio potencial de las stock options.

In [21]:
df['yearswithcurrmanager'].value_counts()

yearswithcurrmanager
2     395
0     285
7     275
3     152
8     122
4     108
1      87
9      71
5      39
10     33
6      31
11     22
12     20
13     17
17      8
15      5
14      5
16      3
Name: count, dtype: int64

In [22]:
df['remotework'].unique()

array(['Yes', '1', 'False', '0', 'True'], dtype=object)

In [51]:
df['maritalstatus'].sample(4)

105     Married
290      Single
1668        NaN
782     Married
Name: maritalstatus, dtype: object

observamos errores de normalizacion tanto en unificar minúsculas como en unir valores unique: 2 valores para mismos conjuntos (Divorced y divorced) o (Married y 'Marreid'), lo que nos dispara los nulos (675) como Nan(tasa muy alta 40.23%).
Opcion limpieza nulos sustituir por 'Unknown'. 

In [23]:
df['maritalstatus'].unique()

array([nan, 'Married', 'Divorced', 'Single', 'divorced', 'Marreid'],
      dtype=object)

In [24]:
df['maritalstatus'].value_counts()

maritalstatus
Married     419
Single      343
Divorced    194
Marreid      36
divorced     11
Name: count, dtype: int64

In [25]:
df['maritalstatus'].isnull().sum()

np.int64(675)

In [26]:
tasa_estado_civil = df['maritalstatus'].value_counts(normalize=True, dropna=False) * 100 #OBTENEMOS PORCENTAJE MANTENIENDO NULOS 
tasa_estado_civil

maritalstatus
NaN         40.226460
Married     24.970203
Single      20.441001
Divorced    11.561383
Marreid      2.145411
divorced     0.655542
Name: proportion, dtype: float64

CORRECCIÓN DE LOS DATOS:
1. SUBSCRIBIMOS COLUMNA
2. limpiamos valores columna a minúsculas
3. modificamos valor NaN por 'unknown', no remplazamos nulos sino que los mantenemos con una categoría especial, dado que es superior a la categoría dominante married.
4. remplazar 'marreid' por 'married'

In [50]:
df['distancefromhome'].sample(4)

1444     2
998     25
1472     1
183    -37
Name: distancefromhome, dtype: int64

observamos valores negativos NaN en la columna "distancefromhome"

In [ ]:
#¿QUE CANTIDAD DE VALORES NEGATIVOS TENEMOS?
numerador = df[df['distancefromhome']<0].shape[0]
print(f'numerador: {numerador}')
denominador = df.shape[0] #accedemos al numero total de filas
print(f'denominador: {denominador}')
porcentaje_negativos = round((numerador/denominador)*100,2)#evitamos números mágicos
print(f'porcentaje_negativos: {porcentaje_negativos}')

numerador: 202
denominador: 1678
porcentaje_negativos: 12.04


In [40]:
lista_unicos = df['distancefromhome'].unique().tolist()
lista_unicos.sort()
print(lista_unicos) #OBSERVAMOS QUE SON NUMEROS NEGATIVOS DE [-49:-10] Y POSITIVOS DE [1:29]
# NO SE REPITEN VALORES ÚNICOS CON DISTINTO VALOR

[-49, -48, -47, -46, -45, -44, -43, -42, -41, -40, -39, -38, -37, -36, -35, -34, -33, -32, -31, -30, -29, -28, -27, -26, -25, -24, -23, -22, -21, -20, -19, -18, -17, -16, -15, -14, -13, -12, -11, -10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]


In [ ]:
filas_negativos = df[df['distancefromhome']<0].head() #FILTRAMOS NEGATIVOS
filas_negativos

age attrition businesstravel    dailyrate department  distancefromhome  \
18  35        No            NaN   556.256661        NaN               -13   
24  36        No  travel_rarely   639.785714        NaN               -42   
30  41        No            NaN  1077.865079     Sales                -37   
41  41        No  travel_rarely   487.087302        NaN               -35   
57  49        No            NaN   556.256661        NaN               -26   

    education educationfield  employeecount  employeenumber  \
18          5  Life Sciences              1              19   
24          2  Life Sciences              1              25   
30          1            NaN              1              31   
41          3            NaN              1              42   
57          4        Medical              1              58   

    environmentsatisfaction  gender  hourlyrate  jobinvolvement  joblevel  \
18                        2       0         NaN               4         2   
24                       42       0   79.973214               3         2   
30                       37       1         NaN               3         3   
41                       35       0   60.885913               3         1   
57                        3       1   69.532083               3         2   

                     jobrole  jobsatisfaction maritalstatus monthlyincome  \
18       REseArcH sciENtiST                 1       Married      4492,84$   
24          SALEs exeCUtIVe                 4           NaN      5167,50$   
30          Sales ExECuTIVe                 3      Divorced      8705,83$   
41    LaboRAtorY TecHNiCiAn                 1        Single      3934,17$   
57   maNufActUrInG DIrECtOr                 1           NaN           NaN   

   monthlyrate  numcompaniesworked over18 overtime  percentsalaryhike  \
18   11681,39$                   1    NaN      NaN                 11   
24   13435,50$                   1      Y      NaN                 14   
30   22635,17$                   0    NaN      NaN                 13   
41   10228,83$                   2    NaN      Yes                 13   
57   11681,39$                   1    NaN       No                 14   

   performancerating  relationshipsatisfaction standardhours  \
18               3,0                         4           NaN   
24               3,0                         4     Full Time   
30               3,0                         4     Full Time   
41               NaN                         3     Full Time   
57               3,0                         3     Part Time   

    stockoptionlevel totalworkingyears  trainingtimeslastyear worklifebalance  \
18                 0               NaN                      2             3,0   
24                 1              18,0                      1             NaN   
30                 1               NaN                      3             4,0   
41                 0              20,0                      3             3,0   
57                 0              16,0                      2             2,0   

    yearsatcompany yearsincurrentrole  yearssincelastpromotion  \
18              16                NaN                        1   
24              18                NaN                        4   
30              22                NaN                       13   
41              18                NaN                        2   
57              15                NaN                        5   

    yearswithcurrmanager sameasmonthlyincome  datebirth      salary  \
18                    10            4492,84$       1988         NaN   
24                    11            5167,50$       1987   62010,00$   
30                     5            8705,83$       1982  104470,00$   
41                    17            3934,17$       1982   47210,00$   
57                    11                 NaN       1974   53914,11$   

                roledepartament  numberchildren remotework  
18                          NaN        

NOTAS ESTRATEGIAS:

1. OPCIÓN1 <Corrección por Valor Absoluto>:
podemos asumir que el signo negativo se trata de un error de entrada de la información. solución, convertir todos los valores negativos a su valor absoluto positivo. Adecuada si el número de casos es pequeño.
2. OPCIÓN2 <Imputación por mediana o media>:
sería una opción si consideramos que la calidad del dato es cuestionable. La mediana es más robusta y clave pera la distancia.
3. OPCIÓN3 <Eliminación Filas>:
no contemplamos esta opción, no queremos arriesgar el análisis. Puede impedir entender el conjunto de datos o romper la continuidad del índice.

CONCLUSIÓN:

Dado que el porcentaje de negativos es 12%, y teniendo en cuenta que es un porcentaje alto, preservamos la integridad de la información original y elegimos la opción1. Corregimos signo, convirtiendo a Valor absoluto, ya la imputación significaría un 12% de los datos.

### MODIFICACIONES: pasar columnas object a float

In [ ]:
df.select_dtypes(include=['object']).columns

Index(['age', 'attrition', 'businesstravel', 'department', 'educationfield',
       'jobrole', 'maritalstatus', 'monthlyincome', 'monthlyrate', 'over18',
       'overtime', 'performancerating', 'standardhours', 'totalworkingyears',
       'worklifebalance', 'yearsincurrentrole', 'sameasmonthlyincome',
       'salary', 'roledepartament', 'remotework'],
      dtype='object')


Existen columnas que son tipo obect, pero que deberían ser float: monthlyincome, monthlyrate, totalworkingyears, salary

NOTAS ESTRATEGIAS: si no se convierten a float los resultados son engañosos.
 
Crear bucle o función para pasar las columnas: 
- Se debe eliminar el signo $ 
- Reemplazar las comas (,) por puntos (.). 
- Convertir a float

In [ ]:
columnas_a_float = ["monthlyincome", "monthlyrate", "totalworkingyears", "salary"]
print(df[columnas_a_float].dtypes)